File: verba-auth/src/main/java/ru/nsu/ccfit/verba/verbaauth/dto/Response.java
package ru.nsu.ccfit.verba.verbaauth.dto;


import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Value;
import ru.nsu.ccfit.verba.verbaauth.enumeration.Code;

@Value
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class Response<T> {

    Code code;

    String message;

    T data;

    public static <T> Response<T> withData(T data) {
        return new Response<>(Code.OK, null, data);
    }

    public static <T> Response<T> withError(Code code, String errorMessage) {
        return new Response<>(code, errorMessage, null);
    }

    public static Response<Void> withoutErrors() {
        return new Response<>(Code.OK, null, null);
    }
}

File: verba-api/src/main/kotlin/ru/nsu/ccfit/verba/verbaapi/platform/Response.kt
package ru.nsu.ccfit.verba.verbaapi.platform

class Response<T>(
    var code: Code,
    var message: String?,
    var data: T?
) {


    companion object {
        fun <T> withData(data: T): Response<T> {
            return Response(Code.OK, null, data)
        }

        fun <T> withError(code: Code, errorMessage: String?): Response<T> {
            return Response(code, errorMessage, null)
        }

        fun withoutErrors(): Response<Void> {
            return Response(Code.OK, null, null)
        }
    }
}
